### PR TITLE
`IRP_MJ_READ`, `IRP_MJ_WRITE`, and `IRP_MJ_SET_INFORMATION`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Continuation of https://github.com/gravitational/teleport/pull/13398

The base branch of https://github.com/gravitational/teleport/pull/13398 had changed enough that it was simpler to just manually rebase (cherry-pick) and open a new PR. See https://github.com/gravitational/teleport/pull/13398 for the previous discussion.

closes https://github.com/gravitational/teleport/issues/14028

### CODE REVIEW

**Review by ibeckermayer — COMMENTED** [review]
Magic!
image


I also added a couple of related issues that I came up with while reviewing:
- https://github.com/gravitational/teleport/issues/14029
- https://github.com/gravitational/teleport/issues/14028

**Review by ibeckermayer — COMMENTED** [review]
I can't officially approve because I created the PR, but approved.

File: lib/srv/desktop/rdp/rdpclient/client.go [hunk 1]
```diff
@@ -468,6 +468,37 @@ func (c *Client) start() {
...
 				}
+			case tdp.SharedDirectoryReadResponse:
+				if c.cfg.AllowDirectorySharing {
+
+					var readData *C.uint8_t
+					if m.ReadDataLength > 0 {
+						readData = (*C.uint8_t)(unsafe.Pointer(&m.ReadData[0]))
+					} else {
+						readData = (*C.uint8_t)(unsafe.Pointer(&m.ReadData))
+					}
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/client.go:479]: Is the Rust code going to read this data or write to this memory?

File: lib/srv/desktop/tdp/proto.go [hunk 2]
```diff
@@ -1102,6 +1114,258 @@ func decodeSharedDirectoryListResponse(in peekReader) (SharedDirectoryListRespon
...
+	Length       uint32
+}
+
+func (s SharedDirectoryReadRequest) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	buf.WriteByte(byte(TypeSharedDirectoryReadRequest))
+	binary.Write(buf, binary.BigEndian, s.CompletionID)
+	binary.Write(buf, binary.BigEndian, s.DirectoryID)
+	// binary.Write(buf, binary.BigEndian, s.PathLength)
```
Comment (zmb3) [inline lib/srv/desktop/tdp/proto.go:1134]: Why is this commented out?

**Review by zmb3 — APPROVED** [review]

File: lib/srv/desktop/tdp/proto.go (hunk 2 above)
Comment (LKozlowski) [reply lib/srv/desktop/tdp/proto.go:1134]: The commented code should be removed. At first I wrote path length + path string in two separate calls, but then found out that we have a helper function that does the same and I've changed it. In the next line you can see that I'm using `encodeString` function.

File: lib/srv/desktop/rdp/rdpclient/client.go (hunk 1 above)
Comment (LKozlowski) [reply lib/srv/desktop/rdp/rdpclient/client.go:479]: Rust code is reading this data. This is handler of response for a read call on the Windows side, so we'll read from it and pass the data into Windows using RDP.

**Pull request merged**